memcached.c

```diff
@@ -4191,6 +4287,837 @@ static inline void process_get_command(conn *c, token_t *tokens, size_t ntokens,
+            fp--; // walk backwards for next token.
+            it->it_flags |= ITEM_TOKEN_SENT;
+        }
+
+        // TODO: Benchmark unlocking here vs later. _get_extstore() could be
+        // intensive so probably better to unlock here after we're done
+        // fiddling with the item header.
+
+        *p = '\r';
+        *(p+1) = '\n';
+        *(p+2) = '\0';
+        p += 2;
+        // tag initial space to the front of the buffer, ahead of any extra
+        // flags that were added.
+        *fp = ' ';
+        *(fp-1) = 'A';
+        *(fp-2) = 'V';
+        fp -= 2;
+        // finally, chain in the buffer.
+        // fp includes the flags.
+        add_iov(c, fp, p - fp);
+
+        if (of.value) {
+#ifdef EXTSTORE
+            if (it->it_flags & ITEM_HDR) {
+                // this bizarre interface is instructing _get_extstore() to
+                // "walk back and zero out" this many iovec's on an internal
+                // miss. kills the VALUE + key + header stitched above.
+                int iovcnt = 2;
+                int iovst = c->iovused - 1;
+
+                if (_get_extstore(c, it, iovst, iovcnt) != 0) {
+                    pthread_mutex_lock(&c->thread->stats.mutex);
+                    c->thread->stats.get_oom_extstore++;
+                    pthread_mutex_unlock(&c->thread->stats.mutex);
+
+                    failed = true;
+                }
+            } else if ((it->it_flags & ITEM_CHUNKED) == 0) {
+                add_iov(c, ITEM_data(it), it->nbytes);
+            } else {
+                add_chunked_item_iovs(c, it, it->nbytes);
+            }
+#else
+            if ((it->it_flags & ITEM_CHUNKED) == 0) {
+                add_iov(c, ITEM_data(it), it->nbytes);
+            } else {
+                add_chunked_item_iovs(c, it, it->nbytes);
+            }
+#endif
+        }
+
+        // FIXME: is extstore counting this? I don't believe it is.
+        if (!c->noreply) {
+            add_iov(c, "EN\r\n", 4);
+        }
+
+        // need to hold the ref at least because of the key above.
+#ifdef EXTSTORE
+        if (!failed) {
+            if ((it->it_flags & ITEM_HDR) != 0 && of.value) {
+                // Only have extstore clean if header and returning value.
+                c->item = NULL;
+            } else {
+                c->item = it;
+            }
+        } else {
+            if (of.locked) {
+                do_item_remove(it);
+            } else {
+                item_remove(it);
+            }
+        }
+#else
+        c->item = it;
+#endif
+    } else {
+        failed = true;
+    }
+
+    if (of.locked) {
+        // Delayed bump so we could get fetched/last access time/etc.
+        // TODO: before segmented LRU, last-access time would only update
+        // every 60 seconds. Currently it doesn't update at all if an item is
+        // marked as ACTIVE. I believe this is a bug... in segmented mode
+        // there's no reason to avoid bumping la on every access.
```

Comment:
+1. So `mg <key> uv` is the only way to get the value without bumping the LRU position? If the ascending order of `la` values indicates the LRU order, we can use that to test the `u` flag.

Reply:
The hit marker is accurate, so I used that to verify u in the tests. That's another way to check it too.